cmd/lotus-shed/fip-0036.go

```diff
@@ -0,0 +1,565 @@
+					m = append(m, addr)
+					workerMap[workerId] = m
+				} else {
+					n := []address.Address{addr}
+					workerMap[workerId] = n
+				}
+
+				bal, err := m.AvailableBalance(act.Balance)
+				if err != nil {
+					return err
+				}
+
+				pow, ok, err := powerState.MinerPower(addr)
+				if err != nil {
+					return err
+				}
+
+				if !ok {
+					pow.RawBytePower = big.Zero()
+				}
+
+				minerActorsInfo[addr] = minerBriefInfo{
+					rawBytePower: pow.RawBytePower,
+					// gets added up outside this loop
+					dealPower: big.Zero(),
+					balance:   bal,
+				}
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			return err
+		}
+
+		fmt.Println("iterating over proposals")
+		dps, _ := marketState.Proposals()
+		if err := dps.ForEach(func(dealID abi.DealID, d market.DealProposal) error {
+			clientId := lookupId(d.Client)
+			if cd, found := clientToDealStorage[clientId]; found {
+				clientToDealStorage[clientId] = big.Add(cd, big.NewInt(int64(d.PieceSize)))
+			} else {
+				clientToDealStorage[clientId] = big.NewInt(int64(d.PieceSize))
+			}
+
+			providerId := lookupId(d.Provider)
+			mai, found := minerActorsInfo[providerId]
+
+			if !found {
+				return xerrors.Errorf("didn't find miner %s", providerId)
+			}
+
+			mai.dealPower = big.Add(mai.dealPower, big.NewInt(int64(d.PieceSize)))
+			minerActorsInfo[providerId] = mai
+			return nil
+		}); err != nil {
+			return xerrors.Errorf("fail to get deals")
+		}
+
+		// now tabulate votes
+
+		approveBalance := abi.NewTokenAmount(0)
+		rejectionBalance := abi.NewTokenAmount(0)
+		clientApproveBytes := big.Zero()
+		clientRejectBytes := big.Zero()
+		msigPendingVotes := make(map[address.Address]msigVote) //map[msig ID]msigVote
+		votedMsigs := make(map[address.Address]struct{})
+		votesIncludingMsigs := make(map[address.Address]uint64)
+		fmt.Println("counting account and multisig votes")
+		for signer, v := range votes {
+			signerId := lookupId(signer)
+			//process votes for regular accounts
+			accountActor, err := st.GetActor(signerId)
+			if err != nil {
+				return xerrors.Errorf("fail to get account account for signer: ", err)
+			}
+
+			clientBytes, ok := clientToDealStorage[signerId]
+			if !ok {
+				clientBytes = big.Zero()
+			}
+
+			if v == APPROVE {
+				approveBalance = types.BigAdd(approveBalance, accountActor.Balance)
+				votesIncludingMsigs[signerId] = APPROVE
+				clientApproveBytes = big.Add(clientApproveBytes, clientBytes)
+			} else {
+				rejectionBalance = types.BigAdd(rejectionBalance, accountActor.Balance)
+				votesIncludingMsigs[signerId] = Reject
+				clientRejectBytes = big.Add(clientRejectBytes, clientBytes)
+			}
+
+			//process msigs
+			// TODO: Oh god, oh god, there's a possibility that a multisig has voted in both directions
+			// and we'll pick a winner non-deterministically as we iterate...
+			// We need to factor in vote time if that happens and pick whoever went first
```

Comment:
(IMO, just pick the direction with more votes and drop the vote on tie)

Reply:
I do think it's correcter to pick the first vote that crossed the threshold -- that's how msigs operate on chain. 

(gonna do that for now, and log how many instances of this we have... @ribasushi already has this info too).

Reply:
@arajasek on my side I am not coming up with any instances of this, aside from the mega-msig

Reply:
I don't agree that it's correcter to pick the first vote that crosses the threshold. That incentivizes a race as soon as the poll opens. We should want people to take as much time as they need before the poll closes to gather and weigh information. 

I agree with @Stebalien's idea as the most straightforward solution.